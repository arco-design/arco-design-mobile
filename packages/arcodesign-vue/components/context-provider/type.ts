export interface ContextInstallOptions {
    /**
     * 组件名称前缀
     * @en Component prefix
     * @default "Arco"
     */
    componentPrefix?: string;
}

export interface GlobalContextParams {
    /**
     * 组件类名前缀
     * @en Component classname prefix
     * @default "arco"
     */
    prefixCls?: string;
}
