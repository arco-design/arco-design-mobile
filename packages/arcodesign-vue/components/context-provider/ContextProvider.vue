<script lang="ts" setup>
import { provide } from 'vue';
import { GlobalContextParams } from './type';
import { configProviderInjectionKey, defaultContext } from './utils';

const props = withDefaults(defineProps<GlobalContextParams>(), defaultContext);
provide(configProviderInjectionKey, props);
</script>
<template>
    <slot />
</template>
