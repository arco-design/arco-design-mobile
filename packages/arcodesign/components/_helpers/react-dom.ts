import { ReactElement } from 'react';
import * as ReactDOM from 'react-dom';

function isObject(obj: any): obj is { [key: string]: any } {
    return Object.prototype.toString.call(obj) === '[object Object]';
}

export interface RootType {
    render: (element: ReactElement) => void;
    _unmount: () => void;
}
export interface RootTypeReact extends RootType {
    unmount?: () => void;
}
export type CreateRootFnType = (container: Element | DocumentFragment) => {
    render: (element: ReactElement) => void;
    unmount?: () => void;
};

// Cast ReactDOM to a version that might have createRoot and other properties
const typedReactDOM = ReactDOM as typeof ReactDOM & {
    createRoot?: CreateRootFnType;
    render?: (
        element: ReactElement,
        container: Element | DocumentFragment,
        callback?: () => void,
    ) => any;
    unmountComponentAtNode?: (container: Element | DocumentFragment) => boolean;
    __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED?: {
        usingClientEntryPoint?: boolean;
    };
};

let copyRender: (
    app: ReactElement,
    container: Element | DocumentFragment,
    createRootFunction?: CreateRootFnType,
) => RootType;

const updateUsingClientEntryPoint = (skipWarning?: boolean) => {
    // https://github.com/facebook/react/blob/17806594cc28284fe195f918e8d77de3516848ec/packages/react-dom/npm/client.js#L10
    // Avoid console warning
    const secretInternals = typedReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    if (isObject(secretInternals)) {
        secretInternals.usingClientEntryPoint = skipWarning;
    }
};

const createRootFn = typedReactDOM.createRoot;

const getRender =
    (createRootFunction: CreateRootFnType) =>
    (app: ReactElement, container: Element | DocumentFragment) => {
        updateUsingClientEntryPoint(true);
        const root = createRootFunction!(container);
        updateUsingClientEntryPoint(false);

        root.render(app);

        return {
            render: (elementToRender: ReactElement) => {
                root.render(elementToRender);
            },
            _unmount() {
                setTimeout(() => {
                    if (root && typeof root.unmount === 'function') {
                        root.unmount();
                    }
                });
            },
        };
    };

if (createRootFn) {
    // React 18
    copyRender = getRender(createRootFn);
} else if (
    typeof typedReactDOM.render === 'function' &&
    typeof typedReactDOM.unmountComponentAtNode === 'function'
) {
    // React 16/17
    copyRender = function (app: ReactElement, container: Element | DocumentFragment) {
        typedReactDOM.render!(app, container); // Use non-null assertion
        return {
            render: (elementToRender: ReactElement) => {
                typedReactDOM.render!(elementToRender, container); // Use non-null assertion
            },
            _unmount() {
                typedReactDOM.unmountComponentAtNode!(container); // Use non-null assertion
            },
        };
    };
} else {
    copyRender = (
        app: ReactElement,
        container: Element | DocumentFragment,
        createRootFunction?: CreateRootFnType,
    ) => {
        const defaultCb = () => {
            // Fallback if no rendering method is found
            console.error(
                'ArcoDesign: ReactDOM.createRoot() or ReactDOM.render() not found. ' +
                    'This usually means you are using an unsupported version of React, ' +
                    'or ReactDOM is not properly initialized. ' +
                    'ArcoDesign Mobile React requires React 16, 17, 18, or 19.',
            );
            return { render: (_element: ReactElement) => {}, _unmount: () => {} };
        };
        if (createRootFunction) {
            return getRender(createRootFunction)(app, container);
        }
        return defaultCb();
    };
}

export const render = copyRender;
