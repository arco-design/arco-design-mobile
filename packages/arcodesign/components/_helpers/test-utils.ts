import React, { useEffect, useRef } from 'react';
import { Promise as ES6Promise, Promise } from 'es6-promise';

/**
 * React 19 兼容的 act 包装器
 * 用于包装可能触发状态更新的异步操作
 */
export const actWrapper = (callback: () => void | ES6Promise<void>): ES6Promise<void> => {
    return new ES6Promise<void>((resolve, reject) => {
        try {
            // 尝试使用 react-dom/test-utils 的 act
            const testUtils = require('react-dom/test-utils');
            if (testUtils && typeof testUtils.act === 'function') {
                // 对于同步回调
                if (typeof callback === 'function') {
                    try {
                        const result = callback();
                        if (result && typeof result.then === 'function') {
                            testUtils
                                .act(() => result)
                                .then(() => resolve())
                                .catch(reject);
                        } else {
                            testUtils.act(() => {
                                callback();
                            });
                            resolve();
                        }
                    } catch (error) {
                        reject(error);
                    }
                } else {
                    resolve();
                }
            } else {
                const result = callback();
                if (result && typeof result.then === 'function') {
                    result.then(() => resolve()).catch(reject);
                } else {
                    resolve();
                }
            }
        } catch (error) {
            // 如果 act 不可用，直接执行回调
            try {
                const result = callback();
                if (result && typeof result.then === 'function') {
                    result.then(() => resolve()).catch(reject);
                } else {
                    resolve();
                }
            } catch (callbackError) {
                reject(callbackError);
            }
        }
    });
};

/**
 * 清理未处理的异步操作，防止内存泄漏警告
 */
export const cleanupAsync = (): ES6Promise<void> => {
    return new ES6Promise<void>(resolve => {
        setTimeout(() => resolve(), 0);
    });
};

/**
 * 安全的状态更新，检查组件是否已卸载
 */
export const createSafeStateUpdater = (isMountedRef: React.MutableRefObject<boolean>) => {
    return (updater: () => void) => {
        if (isMountedRef.current) {
            updater();
        }
    };
};

/**
 * 用于测试的 useEffect 清理函数
 */
export const useTestCleanup = (cleanupFn: () => void) => {
    const isMountedRef = useRef(true);

    useEffect(() => {
        return () => {
            isMountedRef.current = false;
            cleanupFn();
        };
    }, [cleanupFn]);

    return isMountedRef;
};
