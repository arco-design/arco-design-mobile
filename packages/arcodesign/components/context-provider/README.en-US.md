### Others

# ContextProvider 

Global data control component, used to replace global data.

======

> Props

|Property|Description|Type|DefaultValue|
|----------|-------------|------|------|
|prefixCls|Component classname prefix|string|"arco"|
|system|Manually control the current system, and the incoming value will be used directly after being passed in\. It is applicable when the initial value of the system needs to be specified in the ssr scenario\.|"" \| "pc" \| "android" \| "ios"|""|
|useDarkMode|Whether to use dark mode|boolean|false|
|isDarkMode|Whether it is in dark mode|boolean|false|
|theme|Theme variable\. The css variable will be replaced online after input\. The less variable needs to be set|Record\<string, string\>|-|
|locale|Internationalized language configuration|ILocale|-|
|useRtl|Whether to use rtl|boolean|false|

> ILocale

|Property|Description|Type|
|----------|-------------|------|
|locale|Language Type|string|
|LoadMore|-|\{ loadMoreText: string; loadingText: string; prepareText: string; noDataText: string; failLoadText: string; prepareScrollText: string; prepareClickText: string; \}|
|Picker|-|\{ okText: string; cancelText: string; \}|
|Tag|-|\{ addTag: string; \}|
|Dialog|-|\{ okText: string; cancelText: string; \}|
|SwipeLoad|-|\{ normalText: string; activeText: string; \}|
|PullRefresh|-|\{ loadingText: string; pullingText: string; finishText: string; loosingText: string; \}|
|DropdownMenu|-|\{ select: string; \}|
|Pagination|-|\{ previousPage: string; nextPage: string; \}|
|Image|-|\{ loadError: string; \}|
|ImagePicker|-|\{ loadError: string; \}|
|SearchBar|-|\{ placeholder: string; cancelBtn: string; \}|
|Stepper|-|\{ minusButtonName: string; addButtonName: string; \}|
|Keyboard|-|\{ confirm: string; \}|
|Form|-|\{ required: string; type: \{ email: string; url: string; string: string; number: string; array: string; object: string; boolean: string; \}; number: \{ min: string; max: string; equal: string; range: string; positive: string; negative: string; \}; string: \{ \.\.\.; \}; array: \{ \.\.\.; \}; object: \{ \.\.\.; \}; boolean: \{ \.\.\.; \}; \}|
