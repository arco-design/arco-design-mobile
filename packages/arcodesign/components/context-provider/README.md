### 其他

# 全局配置 ContextProvider

全局数据控制组件，用于替换全局数据。

======

> 属性/Props

|参数|描述|类型|默认值|
|----------|-------------|------|------|
|prefixCls|组件类名前缀|string|"arco"|
|system|手动控制当前所在系统，传入后将直接使用传入的值，ssr场景需指定系统初始值时适用|"" \| "pc" \| "android" \| "ios"|""|
|useDarkMode|是否使用暗黑模式|boolean|false|
|isDarkMode|是否处于暗黑模式，指定后以指定的值为准|boolean|false|
|theme|主题变量，传入后将在线替换css变量，需设置less变量 @use\-css\-vars: 1|Record\<string, string\>|-|
|locale|国际化语言包配置|ILocale|-|
|useRtl|是否使用Rtl模式|boolean|false|

> ILocale

|参数|描述|类型|
|----------|-------------|------|
|locale|语言类型|string|
|LoadMore|-|\{ loadMoreText: string; loadingText: string; prepareText: string; noDataText: string; failLoadText: string; prepareScrollText: string; prepareClickText: string; \}|
|Picker|-|\{ okText: string; cancelText: string; \}|
|Tag|-|\{ addTag: string; \}|
|Dialog|-|\{ okText: string; cancelText: string; \}|
|SwipeLoad|-|\{ normalText: string; activeText: string; \}|
|PullRefresh|-|\{ loadingText: string; pullingText: string; finishText: string; loosingText: string; \}|
|DropdownMenu|-|\{ select: string; \}|
|Pagination|-|\{ previousPage: string; nextPage: string; \}|
|Image|-|\{ loadError: string; \}|
|ImagePicker|-|\{ loadError: string; \}|
|SearchBar|-|\{ placeholder: string; cancelBtn: string; \}|
|Stepper|-|\{ minusButtonName: string; addButtonName: string; \}|
|Keyboard|-|\{ confirm: string; \}|
|Form|-|\{ required: string; type: \{ email: string; url: string; string: string; number: string; array: string; object: string; boolean: string; \}; number: \{ min: string; max: string; equal: string; range: string; positive: string; negative: string; \}; string: \{ \.\.\.; \}; array: \{ \.\.\.; \}; object: \{ \.\.\.; \}; boolean: \{ \.\.\.; \}; \}|
