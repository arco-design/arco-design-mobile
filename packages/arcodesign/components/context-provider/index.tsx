import React, { createContext, useState, useEffect, useCallback } from 'react';
import { addCssRules, removeCssStyleDom, defaultLocale, ILocale } from '@arco-design/mobile-utils';

export interface GlobalContextParams {
    /**
     * 组件类名前缀
     * @en Component classname prefix
     * @default "arco"
     */
    prefixCls?: string;
    /**
     * 手动控制当前所在系统，传入后将直接使用传入的值，ssr场景需指定系统初始值时适用
     * @en Manually control the current system, and the incoming value will be used directly after being passed in. It is applicable when the initial value of the system needs to be specified in the ssr scenario.
     * @default ""
     */
    system?: 'pc' | 'android' | 'ios' | '';
    /**
     * 是否使用暗黑模式
     * @en Whether to use dark mode
     * @default false
     */
    useDarkMode?: boolean;
    /**
     * 是否处于暗黑模式，指定后以指定的值为准
     * @en Whether it is in dark mode
     * @default false
     */
    isDarkMode?: boolean;
    /**
     * 主题变量，传入后将在线替换css变量，需设置less变量 @use-css-vars: 1
     * @en Theme variable. The css variable will be replaced online after input. The less variable needs to be set @use-css-vars: 1
     */
    theme?: Record<string, string>;
    /**
     * 国际化语言包配置
     * @en Internationalized language configuration
     */
    locale?: ILocale;
    /**
     * 是否使用Rtl模式
     * @en Whether to use rtl
     * @default false
     */
    useRtl?: boolean;
}

export const defaultContext: GlobalContextParams = {
    prefixCls: 'arco',
    system: '',
    useDarkMode: false,
    isDarkMode: false,
    locale: defaultLocale,
    useRtl: false,
};

export const GlobalContext = createContext<GlobalContextParams>(defaultContext);

export interface ContextProviderProps extends GlobalContextParams {
    children: React.ReactNode;
}

export type WithGlobalContext<T> = T & { context?: GlobalContextParams };

/**
 * 全局数据控制组件，用于替换全局数据。
 * @en Global data control component, used to replace global data.
 * @type 其他
 * @type_en Others
 * @name 全局配置
 * @name_en ContextProvider
 */
export default function ContextProvider(props: ContextProviderProps) {
    const {
        children,
        useDarkMode,
        isDarkMode,
        theme,
        locale = defaultLocale,
        ...restProps
    } = props;
    const [isDarkModeState, setIsDarkModeState] = useState<boolean>();
    const changeDarkMode = useCallback(
        (res: MediaQueryListEvent) => setIsDarkModeState(res.matches),
        [],
    );

    useEffect(() => {
        const matchMedia = window.matchMedia('(prefers-color-scheme: dark)');
        if (useDarkMode) {
            const dark = matchMedia.matches;
            setIsDarkModeState(dark);
            if (typeof matchMedia.addEventListener === 'function') {
                matchMedia.addEventListener('change', changeDarkMode);
            } else if (typeof matchMedia.addListener === 'function') {
                matchMedia.addListener(changeDarkMode);
            }
        }
        return () => {
            if (useDarkMode) {
                if (typeof matchMedia.removeEventListener === 'function') {
                    matchMedia.removeEventListener('change', changeDarkMode);
                } else if (typeof matchMedia.removeListener === 'function') {
                    matchMedia.removeListener(changeDarkMode);
                }
            }
        };
    }, [useDarkMode]);

    useEffect(() => {
        if (!theme) {
            return;
        }
        addCssRules('arcoCustomTheme', theme);
        return () => {
            removeCssStyleDom('arcoCustomTheme');
        };
    }, [theme]);

    return (
        <GlobalContext.Provider
            value={{
                ...defaultContext,
                ...restProps,
                isDarkMode: isDarkMode === void 0 ? isDarkModeState : isDarkMode,
                useDarkMode,
                locale: locale || defaultLocale,
            }}
        >
            {children}
        </GlobalContext.Provider>
    );
}

export const ContextLayout = GlobalContext.Consumer;

export function CompWithGlobalContext<P extends JSX.IntrinsicAttributes>(
    Component: React.FunctionComponent<P>,
) {
    return function (props: WithGlobalContext<P>) {
        const { context: propsContext, ...others } = props;
        return (
            <ContextLayout>
                {context => (
                    <ContextProvider {...{ ...context, ...propsContext }}>
                        <Component {...(others as P)} />
                    </ContextProvider>
                )}
            </ContextLayout>
        );
    };
}
