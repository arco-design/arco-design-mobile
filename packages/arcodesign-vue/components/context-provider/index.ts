import ContextProvider from './ContextProvider.vue';
import { installFactory } from '../_helpers/install';

export type * from './type';
export * from './utils';

const install = installFactory(ContextProvider);

export { ContextProvider };

export default {
    install,
};
