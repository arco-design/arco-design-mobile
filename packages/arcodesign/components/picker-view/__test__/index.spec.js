import React, { createRef } from 'react';
import demoTest from '../../../tests/demoTest';
import mountTest from '../../../tests/mountTest';
import PickerView from '..';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { defaultContext } from '../../context-provider';

const pickerPrefix = `${defaultContext.prefixCls}-picker`;

demoTest('picker-view');

mountTest(PickerView, 'PickerView', {
    data: [
        {
            label: '北京市',
            value: '北京市',
            children: [
                {
                    label: '朝阳区',
                    value: '朝阳区',
                },
                {
                    label: '海淀区',
                    value: '海淀区',
                },
                {
                    label: '东城区',
                    value: '东城区',
                },
                {
                    label: '西城区',
                    value: '西城区',
                },
            ],
        },
    ],
    value: ['北京市', '海淀区'],
});

describe('PickerView', () => {

    beforeEach(() => {
        jest.useFakeTimers();
    });

    afterEach(() => {
        jest.useRealTimers();
    });

    it('should render correctly when set default value as ValueType[][]', () => {
        const data = [
            ['周一', '周二', '周三', '周四', '周五', '周六', '周日'],
            ['上午', '下午', '晚上'],
        ];
        const defaultValue = ['周三', '上午'];
        const component = mount(
            <PickerView data={data} value={defaultValue} cascade={false} />,
        );
        const firstCol = component.find(`.${pickerPrefix}-column`).first();
        expect(firstCol.find(`.${pickerPrefix}-column-item`).length).toBe(7);
        expect(firstCol.find('.selected-neighbor-2').length).toBe(2);
        expect(firstCol.find('.selected-neighbor-1').length).toBe(2);
        expect(firstCol.find('.selected').length).toBe(1);
        expect(firstCol.find('.selected').text()).toBe('周三');
        const secondCol = component.find(`.${pickerPrefix}-column`).last();
        expect(secondCol.find(`.${pickerPrefix}-column-item`).length).toBe(3);
        expect(secondCol.find('.selected-neighbor-2').length).toBe(1);
        expect(secondCol.find('.selected-neighbor-1').length).toBe(1);
        expect(secondCol.find('.selected').length).toBe(1);
        expect(secondCol.find('.selected').text()).toBe('上午');
    });
    it('should render correctly when set default value as PickerData[][]', () => {
        const data = [
            [
                { label: '周一', value: '周一' },
                { label: '周二', value: '周二' },
                { label: '周三', value: '周三' },
                { label: '周四', value: '周四' },
                { label: '周五', value: '周五' },
                { label: '周六', value: '周六' },
                { label: '周日', value: '周日' },
            ],
            [
                { label: '上午', value: '上午' },
                { label: '下午', value: '下午' },
                { label: '晚上', value: '晚上' },
            ],
        ];
        const defaultValue = ['周三', '上午'];
        const component = mount(
            <PickerView data={data} value={defaultValue} cascade={false} />,
        );
        const firstCol = component.find(`.${pickerPrefix}-column`).first();
        expect(firstCol.find(`.${pickerPrefix}-column-item`).length).toBe(7);
        expect(firstCol.find('.selected-neighbor-2').length).toBe(2);
        expect(firstCol.find('.selected-neighbor-1').length).toBe(2);
        expect(firstCol.find('.selected').length).toBe(1);
        expect(firstCol.find('.selected').text()).toBe('周三');
        const secondCol = component.find(`.${pickerPrefix}-column`).last();
        expect(secondCol.find(`.${pickerPrefix}-column-item`).length).toBe(3);
        expect(secondCol.find('.selected-neighbor-2').length).toBe(1);
        expect(secondCol.find('.selected-neighbor-1').length).toBe(1);
        expect(secondCol.find('.selected').length).toBe(1);
        expect(secondCol.find('.selected').text()).toBe('上午');
    });
    it('should callback correctly when un cascade picker view is clicked', () => {
        const mockFn = jest.fn();
        const ref = createRef();
        const data = [
            [
                { label: '周一', value: '周一' },
                { label: '周二', value: '周二' },
                { label: '周三', value: '周三' },
                { label: '周四', value: '周四' },
                { label: '周五', value: '周五' },
                { label: '周六', value: '周六' },
                { label: '周日', value: '周日' },
            ],
            [
                { label: '上午', value: '上午' },
                { label: '下午', value: '下午' },
                { label: '晚上', value: '晚上' },
            ],
        ];
        const defaultValue = ['周三', '上午'];
        const component = mount(
            <PickerView ref={ref} data={data} value={defaultValue} onPickerChange={mockFn} cascade={false} />,
        );
        const { getColumnValue } = ref.current;
        const firstCol = component.find(`.${pickerPrefix}-column`).first();
        const neighborNode = firstCol.find(`.${pickerPrefix}-column-item`).first();
        const valueBeforeClick = getColumnValue();
        expect(valueBeforeClick).toBe('周三');
        neighborNode.simulate('click');
        act(() => {
            jest.advanceTimersByTime(200);
        })
        component.update();
        const valueAfterClick = getColumnValue();
        expect(mockFn).toBeCalled();
        expect(valueAfterClick).toBe('周一');
    });
    it('should render and callback correctly when use cascade picker view click', () => {
        const data = [
            {
                label: '北京市',
                value: '北京市',
                children: [
                    {
                        label: '北京市',
                        value: '北京市',
                        children: [
                            {
                                label: '朝阳区',
                                value: '朝阳区',
                            },
                            {
                                label: '海淀区',
                                value: '海淀区',
                            },
                            {
                                label: '东城区',
                                value: '东城区',
                            },
                            {
                                label: '西城区',
                                value: '西城区',
                            },
                        ],
                    },
                ],
            },
            {
                label: '辽宁省',
                value: '辽宁省',
                children: [
                    {
                        label: '沈阳市',
                        value: '沈阳市',
                        children: [
                            {
                                label: '沈河区',
                                value: '沈河区',
                            },
                            {
                                label: '浑南区',
                                value: '浑南区',
                            },
                            {
                                label: '沈北新区',
                                value: '沈北新区',
                            },
                        ],
                    },
                    {
                        label: '本溪市',
                        value: '本溪市',
                        children: [
                            {
                                label: '溪湖区',
                                value: '溪湖区',
                            },
                            {
                                label: '东明区',
                                value: '东明区',
                            },
                            {
                                label: '桓仁满族自治县',
                                value: '桓仁满族自治县',
                            },
                        ],
                    },
                ],
            },
            {
                label: '云南省',
                value: '云南省',
                children: [
                    {
                        label: '昆明市',
                        value: '昆明市',
                        children: [
                            {
                                label: '五华区',
                                value: '五华区',
                            },
                            {
                                label: '官渡区',
                                value: '官渡区',
                            },
                            {
                                label: '呈贡区',
                                value: '呈贡区',
                            },
                        ],
                    },
                ],
            },
        ];
        const value = ['北京市', '北京市', '海淀区'];
        const ref = createRef();
        const components = mount(<PickerView
            cascade={true}
            ref={ref}
            data={data}
            value={value}
            cols={3}
        />);
        const { getColumnValue, getAllColumnValues } = ref.current;
        const firstCol = components.find(`.${pickerPrefix}-column`).first();
        const firstColLastNode = firstCol.find(`.${pickerPrefix}-column-item`).last();
        const valueBeforeClick = getColumnValue();
        const valuesBeforeClick = getAllColumnValues();
        expect(valueBeforeClick).toBe('北京市');
        expect(valuesBeforeClick).toEqual(['北京市', '北京市', '海淀区']);
        firstColLastNode.simulate('click');
        act(() => {
            jest.advanceTimersByTime(200);
        })
        components.update();
        const valueAfterClick = getColumnValue();
        const valuesAfterClick = getAllColumnValues();
        expect(valueAfterClick).toBe('云南省');
        expect(valuesAfterClick).toEqual(['云南省', '昆明市', '五华区']);
    });
});
