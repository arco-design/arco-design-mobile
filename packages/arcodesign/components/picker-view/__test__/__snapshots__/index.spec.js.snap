// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`picker-view demo test picker-view demo: cascader.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-picker-view all-border-box "
    style="height: 0px;"
  >
    <div
      class="arco-picker-multi"
      style="line-height: 0px;"
    >
      <div
        class="arco-picker-column"
      >
        <div
          aria-disabled="false"
          class="arco-picker-column-item-wrap"
          style="transform: translate3d(0px, 0px, 0px); padding-bottom: 0px; padding-top: 0px;"
        >
          <div
            aria-label="Beijing"
            class="arco-picker-column-item selected"
            style="height: 0px;"
          >
            Beijing
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-1"
            style="height: 0px;"
          >
            Liaoning
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-2"
            style="height: 0px;"
          >
            Yunnan
          </div>
        </div>
      </div>
      <div
        class="arco-picker-column"
      >
        <div
          aria-disabled="false"
          class="arco-picker-column-item-wrap"
          style="transform: translate3d(0px, 0px, 0px); padding-bottom: 0px; padding-top: 0px;"
        >
          <div
            aria-label="Beijing"
            class="arco-picker-column-item selected"
            style="height: 0px;"
          >
            Beijing
          </div>
        </div>
      </div>
      <div
        class="arco-picker-column"
      >
        <div
          aria-disabled="false"
          class="arco-picker-column-item-wrap"
          style="transform: translate3d(0px, 0px, 0px); padding-bottom: 0px; padding-top: 0px;"
        >
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-1"
            style="height: 0px;"
          >
            Chaoyang
          </div>
          <div
            aria-label="Haidian"
            class="arco-picker-column-item selected"
            style="height: 0px;"
          >
            Haidian
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-1"
            style="height: 0px;"
          >
            Dongcheng
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-2"
            style="height: 0px;"
          >
            Xicheng
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-picker-selection"
    >
      <div
        class="arco-picker-selection-mask arco-picker-selection-mask-top"
      />
      <div
        class="arco-picker-selection-bar"
      />
      <div
        class="arco-picker-selection-mask arco-picker-selection-mask-bottom"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`picker-view demo test picker-view demo: index.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-picker-view all-border-box "
    style="height: 0px;"
  >
    <div
      class="arco-picker-multi"
      style="line-height: 0px;"
    >
      <div
        class="arco-picker-column"
      >
        <div
          aria-disabled="false"
          class="arco-picker-column-item-wrap"
          style="transform: translate3d(0px, 0px, 0px); padding-bottom: 0px; padding-top: 0px;"
        >
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-2"
            style="height: 0px;"
          >
            Monday
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-1"
            style="height: 0px;"
          >
            Tuesday
          </div>
          <div
            aria-label="Wednesday"
            class="arco-picker-column-item selected"
            style="height: 0px;"
          >
            Wednesday
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-1"
            style="height: 0px;"
          >
            Thursday
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-2"
            style="height: 0px;"
          >
            Friday
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item"
            style="height: 0px;"
          >
            Saturday
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item"
            style="height: 0px;"
          >
            Sunday
          </div>
        </div>
      </div>
      <div
        class="arco-picker-column"
      >
        <div
          aria-disabled="false"
          class="arco-picker-column-item-wrap"
          style="transform: translate3d(0px, 0px, 0px); padding-bottom: 0px; padding-top: 0px;"
        >
          <div
            aria-label="Morning"
            class="arco-picker-column-item selected"
            style="height: 0px;"
          >
            Morning
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-1"
            style="height: 0px;"
          >
            Afternoon
          </div>
          <div
            aria-label=""
            class="arco-picker-column-item selected-neighbor-2"
            style="height: 0px;"
          >
            Evening
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-picker-selection"
    >
      <div
        class="arco-picker-selection-mask arco-picker-selection-mask-top"
      />
      <div
        class="arco-picker-selection-bar"
      />
      <div
        class="arco-picker-selection-mask arco-picker-selection-mask-bottom"
      />
    </div>
  </div>
</DocumentFragment>
`;
